keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -957,55 +961,89 @@ complete their service.
+predefined objects to minimize their effects. A
+PriorityLevelConfiguration can be set to have just 1 assured
+concurrency share; care should be taken to do this only after its
+queues have drained, otherwise this setting may greatly slow down the
+rate at which the queues drain. A FlowSchema can be updated to match
+no requests; following is an example.
 
-```yaml
+```
 kind: FlowSchema
 meta:
-  name: non-top-backstop
+  name: matches-nothing
 spec:
-  matchingPriority: (doesn’t really matter)
-  requestPriority:
-    name: (name of an effectively existing RequestPriority, whether
-           that is real or backstop, with catchAll==true)
-  flowDistinguisher:
-    source: user
-    # no transformation in this case
-  match:
-  - and: [ ] # match everything
+  matchingPrecedence:...
+  priorityLevelConfiguration: ...
+  rules:
+  - rule:
+      verbs: ['*']
+      apiGroups: ['*']
+      resources: ['*']
+    subjects:
+    - kind: Group
+      name: "system:reserved" # a group name that we presume will never be used
 ```
 
-The other part of the defaulting story concerns making actual API
-objects exist, and it goes as follows.  Whenever there is no actual
-RequestPriority object with `exempt == true`, the RequestPriority
-objects exhibited above are created --- except those with a name
-already in use by an existing RequestPriority object.  Whenever there
-is no actual FlowSchema object that refers to an exempt
-RequestPriority object, the schema objects shown above as examples are
-generated --- except those with a name already in use.
+#### API Server Initial Behavior
+
+As it starts up, an apiserver has to start handling requests before
+the filter discussed here has had a chance to read the persisted
+configuration API objects.  Indeed, that reading will have to be done
+through the very same filter (at least in the first apiserver in a
+cluster)!  The initial request handling also precedes persisting the
+default configuration API objects for the cluster.
+
+When the apiserver first starts up, the filter behaves as if the
+configuration consists of just (1) the predefined
+PriorityLevelConfiguration objects named `exempt` and `default`, and
+(2) the predefined FlowSchema objects named `exempt` and `default`.
+
+A kube-apiserver maintains this initial behavior until it has had a
+chance to read all the persisted configuration API objects and create
+any missing ones (see
+[the logic above](#default-cluster-configuration)).
+
+An aggregated apiserver maintains this initial behavior until it has
+had a chance to read all the persisted configuration API objects.
+
+After the initial behavior period is over, regular behavior begins.
+
+#### Default Request Handling
+
+It is possible that the administrators can modify the configuration
+API objects to create a situation where some requests will not match
+any FlowSchema.  Once the initial behavior period is over, any request
+that does not match a FlowSchema object is rejected.
```

Comment:
what about having a default/catch-all for all those requests that don't match any flow schema, instead of have them all 'rejected', put them all into a low priority queue.

the reasoning is there could be new request / new apis added over time, and the current behavior (max-in-flight limit) is not rejecting them

Reply:
I agree that we should do better.  The KEP stipulates similar handling for this case.  #81003 implements it.

Reply:
.. and it is implemented in https://github.com/kubernetes/kubernetes/pull/82196